### PR TITLE
Fix #301210: Paste Half Duration / Paste Double Duration missing from various context menus

### PR DESCRIPTION
**Pull request by Spire42**
Resolves: [#301210](https://musescore.org/en/node/301210)

Added “Paste Half Duration” and “Paste Double Duration” menu items to the context menus for score elements and measures.

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by MarcSabatella** [comment]
Thanks for the PR!  It looks, though, like this will add the menu items for all score elements, when it isn't actually relevant for most.  You can only paste a range selection if the selection is a range or a note or rest.

But that leads to a bigger concern: it doesn't actually work, if there is no selection, and works incorrectly if there is a selection but it's not where you right-clicked.  The paste will go to the location of the selection, regardless of where you right click.  That's not specifically a problem of this PR, but it does very much limit its usefulness.  See https://musescore.org/en/node/300381 and also https://musescore.org/en/node/300126.  Maybe we need to rethink some of this.

**Comment by Spire42** [comment]
> It looks, though, like this will add the menu items for all score elements, when it isn't actually relevant for most. You can only paste a range selection if the selection is a range or a note or rest.

That's also true for the corresponding items in the *Edit* menu: There are always *Paste Half Duration* and *Paste Double Duration* items, and they're always enabled, even if they aren't relevant to the current selection — and even if there's *no* selection!

> But that leads to a bigger concern: it doesn't actually work, if there is no selection, and works incorrectly if there is a selection but it's not where you right-clicked. The paste will go to the location of the selection, regardless of where you right click.

That's also true for the existing *Paste* and *Swap with Clipboard* items in the context menus: They're always there and enabled, they don't work if there's no selection, and using them always affects the selection, regardless of where you right-click.

Given that these two things you mentioned are preexisting issues, my inclination is to merge in this pull request so that the functionality is where it should be and so that these edge cases can get fixed when they eventually get fixed for the other existing cases.

**Comment by MarcSabatella** [comment]
Regular menus are totally different though - they don't suffer that confusion between what is selected and the thing you just right-clicked.  It's always 100% clear they will act on the selection.  Furthermore, the menus don't change their contents depending the selection.  At most, they change what is grayed out, and yes, we should consider graying out all of the copy/paste options according to the selection.

The context menu creates the potential for real user surprise though.  For instance, select a region, Ctrl+C to copy.  Now right-click a note somewhere else, without first clearing the selection, and try to "past double duration".  It may look like nothing happened at all, but in fact you will have just clobbered your original selection.  Whereas with the regular menu, there would be no way to fool youtrself into thinking the paste was going to go anywhere but right there, since it's obvious that you need to select a destination first before pasting.

To me this is a basic flaw in the whole idea of putting things in the context menu that do not in fact have to do with that context, that really should be re-thought.  I'm not sure how many commands suffer this problem, but certainly, the copy/paste commands do pretty much anywhere they are used.

**Comment by Spire42** [comment]
You're essentially arguing that the existing *Paste* and *Swap with Clipboard* items be removed from the content menus, since they're just as broken as the *Paste Half Duration* and *Paste Double Duration* items that are being added in this PR.

I say either put them all in (and fix the broken cases later) or remove them all, but *don't* go halfway and just leave things as they are. And given a choice between putting them all in (adding missing functionality) and removing them all (losing existing functionality), I'd definitely go with the former.

**Comment by MarcSabatella** [comment]
Actual my vote is, first fix what’s broken.  Then consider adding more.

**Comment by Spire42** [comment]
> You can only paste a range selection if the selection is a range or a note or rest.

Could you please create a separate issue for this that includes an exhaustive list of pasting rules like that one? It should be straightforward enough for me or someone else to write the code to disable the pasting menu items as long as we have the actual rules codified.

**Comment by MarcSabatella** [comment]
As far as know, that’s it right there - either range, or a note or rest.  You could check the code for normalPaste() to see if there are any other special cases,  but that’s assuming we’re talking about actually pasting a range selection, which we are in this case.  There might easily be all sorts of other pastes allowed.  Which is probably why no one has thought it worthwhile to micromanage graying out the menu.

**Comment by MarcSabatella** [comment]
But to be clear: disabling the items in the main menu is not the problem I am saying needs fixing; that to me is inconsequential.  The problem is the fact that paste *from the context* menu does not do what the user would expect: it uses the *selection* as the destination, rather than using the *context* (ie, the element that was right-clicked)   To me this makes the inclusion of paste as is a serious bug that needs to be fixed before we compound the problem by adding more menu items that also will not work correctly.  Once we have a general solution to the problem of paste in the context menu, then I think a the paste half/double durations could use the same approach.

**Comment by riaanvn** [comment]
Hi Mark. I agree with the principle of fixing one thing before adding a feature, but these are totally unrelated, the only common denominator is that they are on the context menu, and have to disagree how you applied the principle. 

How will make merging this PR increase the problem? The issue has existed for a long time. We are proposing adding lesser used menu entries to the context menu. If with the added menu entries, people use the existing cut/copy/paste/delete 95% of the time and the paste special entries 5%, how is it making it objectively worse?

Would merging this PR possibly complicate the fix for https://musescore.org/en/node/300381 ?

If we proposed adding a menu entry that would draw a lot of attention to the context menu and the existing right-click bug, then by all means, put the new context menu entries on hold.

Existing users who use the context menus (small number, and possibly dwindling if you are correct), know about the incorrect right-click behaviour, and select with left-click before right-clicking. 

Aside: is there any harm in putting the three paste special entries in Edit and Context menus behind a parent Paste Special menu as part of this PR, or do you think that should be a separate PR after this initial PR to get Edit and context menus in sync?

* Paste Special ->
** Paste Half Duration
** Paste Double Duration
** Swap with Clipboard

**Comment by MarcSabatella** [comment]
They are not unrelated at all.  Paste, as it exists in the right-click menu right now, is broken.  It quite simply does not do what the user should expect.  It pastes the clipboard somewhere other than where he just right-clicked, thus violating the very nature of what a context menu is supposed to be.  If we were to add paste half/double, those commands would be completely broken right from the very beginning in precisely the same way.  We should not add commands that are known to not work.  We should first fix the existing broken ones, then leverage that fix to make sure any new commands we add also work.

To make an analogy: let's say the triplet command was broken, it always corrupted your score, because of a fundamental flaw in how tuplets in general are implemented.  That would not be a good time to then add a quadruplet command that would have the same bug.  You fix the triplet command first, by fixing the fundamental flaw in the tuplet implementation, then leverage that fix to add a quadruplet command that actually works.

I don't think this PR would "complicate" a PR to fix the underlying issue, it would just create more broken code for that PR to need to fix, but presumably the same approach used in fixing the exisitng paste command would fix this also.

I don't really have an opinion on whether to group additional paste commands behind a second level menu.  It makes sense on one level, to simplify the first level menu, but I don't have the sort of UI design background to know what the tradeoffs really are.

**Comment by Spire42** [comment]
The only problem I see with grouping the various paste-related menu items under a “Paste Special” submenu is that MuseScore already has a “Paste Special” command. It doesn't appear under any menus to my knowledge, but it can be assigned a keyboard shortcut.

**Comment by Spire42** [comment]
Marc, you brought up two issues:

1. Some context menu items are not correctly grayed out even if they don't directly apply to the selected item.

2. The wrong context menu pops up if the user right-clicks outside the current selection. (You called this a “bigger concern” than the first item.)

If I'm understanding correctly, your position is that you want this PR be held back from being merged until the above two items have been resolved. Is that correct?

I understand where you're coming from, but I still disagree with you . Would you be willing to compromise and agree to merge this PR once the second item (the “bigger concern”) has been fixed?

**Comment by MarcSabatella** [comment]
My concern is that if the user right-clicks a note then chooses Paste (or, with this PR, Paste Half/Double Duration), the wrong thing happens.  That's unacceptable.  In theory, we could fix it by making sure the menu item does not appear when not valid (eg, when the clicked element is not the beginning of the selection), or whether it is fixed by graying it out in those cases, or whether it is fixed by making sure it actually does what the user has every right to expect it to do (paste to the right-clicked location).  Any of these technically fix the problem.  But the first two are the equivalent of "doctor, it it hurts when I do this / so don't do that".  The third option is best to me.

But once the bug is fixed - by any of the three methods - there is no compromise needed, then will be OK to add the new commands, because they will actually work correctly.  I'm only objecting to adding a command that *doesn't* work, and frankly I don't understand how anyone could possibly support adding a command that doesn't work.  But adding a command that works is good.  So, with the bug fixed, it becomes possible to add the command and make it work.

**Comment by Spire42** [comment]
I didn't quite understand your reply. Let me rephrase my question:

Are you OK with merging this PR once we can fix things so that clicking a context item never does the wrong thing, but sometimes does nothing?

**Comment by MarcSabatella** [comment]
Certainly.  To be clear, I have nothing against adding these commands to the menu.  I am only objecting to adding commands that don't actually work as one would reasonably expect.

**Comment by riaanvn** [comment]
Back to the PR:

1 I found references to these strings in mscore/shortcut.cpp 

do the stanzas for paste-half / paste-double in
mscore/shortcut.cpp
need to be extended? Reason I am asking is that 

         MsWidget::SCORE_TAB,
         STATE_NORMAL,
         "paste-half",

looks different from line two of the equivalent swap stanza (assuming that swap is the clostest to paste-half/-double,

         MsWidget::SCORE_TAB,
         STATE_NORMAL | STATE_NOTE_ENTRY | STATE_EDIT | STATE_TEXT_EDIT | STATE_LYRICS_EDIT
            | STATE_HARMONY_FIGBASS_EDIT,
         "swap",

2 @Spire42 what action does the "Paste Special" keyboard shortcut perform. I thought it would be very rare for an action to have a shortcut way to trigger but not have a menu entry. Since it does not have a menu entry nor is assigned by default, how likely is it that it would be used? 
If Paste Special is a (potential action), would a workaround be to call the menu entry label "Paste Special" but internally have it called something else (e.g. "paste-special-menu") so that it does not clash with the action?

**Comment by MarcSabatella** [comment]
I added the Paste Special command and deliberately didn't put in the menu because there are still questions as to how it should work.  Right now, what it does is allow custom scaling factor instead of just half/double, using the toolbar.  So you can select dotted quarter on the toolbar, then run Paste Special and it multiplies all note values by 1.5.

Eventually there should be a dialog box allowing a more clear way to specify this, and also handling of tuplets for converting between 4/4 and 12/8, etc.  But until then it didn't seem worth exposing to the user.  The command is there because it was practically free to implement and allows for some testing of the functionality, and if it ever comes up in forum, we can tell people about it.

So, now, it's not something we do every day, and I'm not saying I'm particularly proud of doing that here :-)

**Comment by MarcSabatella** [comment]
Swap might be relevant while in text edit mode, but nothing involving durations is.

FWIW, though, swap doesn't seem to actually *work* while in text edit mode.

**Comment by igorkorsukov** [comment]
Unfortunately, now not actual
